pkg/oc/cli/admin/migrate/storage/storage.go

```diff
@@ -267,72 +288,62 @@ func (o MigrateAPIStorageOptions) Run() error {
 		if o.limiter != nil {
 			// we rate limit after performing all operations to make us less sensitive to conflicts
-			// use a defer to make sure we always rate limit after a successful GET even if the PUT fails
-			defer func() {
-				// we have to wait until after the GET to determine how much data we will PUT
-				// thus we need to double the amount to account for both operations
-				// we also need to account for the initial list operation which is roughly another GET per object
-				// thus we can amortize the cost of the list by adding another GET to our calculations
-				// so we have 2 GETs + 1 PUT == 3 * size of data
-				// this is a slight overestimate since every retry attempt will still try to account for
-				// the initial list operation.  this should not be an issue since retries are not that common
-				// and the rate limiting is best effort anyway.  going slightly slower is acceptable.
-				latency := o.limiter.take(3 * len(data))
-				// mimic rest.Request.tryThrottle logging logic
-				if latency > longThrottleLatency {
-					glog.V(4).Infof("Throttling request took %v, request: %s:%s", latency, "GET", r.URL().String())
-				}
-			}()
+			// use a defer to make sure we always rate limit even if the PUT fails
+			defer o.rateLimit(oldObject)
 		}
 
-		update := info.Client.Put().
-			Resource(info.Mapping.Resource.Resource).
-			NamespaceIfScoped(info.Namespace, info.Mapping.Scope.Name() == meta.RESTScopeNameNamespace).
-			Name(info.Name).Body(data).
-			Do()
-		if err := update.Error(); err != nil {
+		// we are relying on unstructured types being lossless and unchanging
+		// across a decode and encode round trip (otherwise this command will mutate data)
+		newObject, err := o.client.
+			Resource(info.Mapping.Resource).
+			Namespace(info.Namespace).
+			Update(oldObject)
+		if err != nil {
 			return migrate.DefaultRetriable(info, err)
 		}
-
-		if oldObject, err := get.Get(); err == nil {
-			info.Refresh(oldObject, true)
-			oldVersion := info.ResourceVersion
-			if object, err := update.Get(); err == nil {
-				info.Refresh(object, true)
-				if info.ResourceVersion == oldVersion {
-					return migrate.ErrUnchanged
-				}
-			} else {
-				glog.V(4).Infof("unable to calculate resource version: %v", err)
-			}
-		} else {
-			glog.V(4).Infof("unable to calculate resource version: %v", err)
+		if newObject.GetResourceVersion() == oldObject.GetResourceVersion() {
+			return migrate.ErrUnchanged
 		}
+	default:
+		return fmt.Errorf("invalid type %T passed to storage migration: %v", oldObject, oldObject)
 	}
 	return nil
 }
 
+func (o *MigrateAPIStorageOptions) rateLimit(oldObject *unstructured.Unstructured) {
+	// we need to approximate how many bytes this object was on the wire
+	// the simplest way to do that is to encode it back into bytes
+	// this is wasteful but we are trying to artificially slow down the worker anyway
+	var dataLen int
+	if data, err := oldObject.MarshalJSON(); err != nil {
```

Comment:
yikes, really?

Reply:
this is really ugly.  the other options are pretty ugly too (reflect walking can get expensive, but is more honest).  him